### PR TITLE
fix(optim): correct SequentialLR initialization when milestones contain 0

### LINKED ISSUE #185872: SequentialLR initializes with the wrong learning rate when milestones contain 0 (due to off by one error)
### 🐛 Describe the bug

# Issue

There's an issue with `SequentialLR(..., milestones=[0, ...])`. \
We use this sometimes to test with no warmup.


- In __init__
    - It [initializes the learning rate with the **FIRST scheduler**, regardless of what the milestones specify](https://github.com/pytorch/pytorch/blob/v2.12.0/torch/optim/lr_scheduler.py#L1168-L1170).
- Then in the step function:
    - `last_epoch` is immediately [incremented to 1](https://github.com/pytorch/pytorch/blob/v2.12.0/torch/optim/lr_scheduler.py#L1187)
    - It then [**fetches the CORRECT scheduler** by running a bisect_right to get the correct idx](https://github.com/pytorch/pytorch/blob/v2.12.0/torch/optim/lr_scheduler.py#L1188-L1189). In this case, `idx` is `1`, so it gets the second scheduler.
    - It should ideally update the learning rate immediately by entering [`if idx > 0 and self._milestones[idx - 1] == self.last_epoch:`](https://github.com/pytorch/pytorch/blob/v2.12.0/torch/optim/lr_scheduler.py#L1190)
    - **[BUG HERE]** But in this case, right after init, `self._milestones[idx - 1]` is `0` and `self.last_epoch` is `1`, so the equality fails, and it goes straight to the else block, and runs [`scheduler.step`](https://github.com/pytorch/pytorch/blob/v2.12.0/torch/optim/lr_scheduler.py#L1193).
    - This runs the correct scheduler, but with the initial learning rate set by the first scheduler.

# Reproducing

## Script
```python
import torch
import torch.optim as optim
from torch import nn
from torch.optim.lr_scheduler import LinearLR, SequentialLR

model = nn.Linear(10, 10)
optimizer = optim.Adam(model.parameters(), lr=1.0)

warmup = LinearLR(optimizer, start_factor=0.1, end_factor=0.5, total_iters=10)
decay = LinearLR(optimizer, start_factor=0.5, end_factor=0.1, total_iters=10)
scheduler = SequentialLR(optimizer, schedulers=[warmup, decay], milestones=[0])

print("Initial LR:", scheduler.get_last_lr()[0])
for _ in range(10):
    scheduler.step()
    print("Step LR:", scheduler.get_last_lr()[0])
```

## Result

- The learning rates should be between 0.5 and 0.1, but instead are between 0.05 and 0.01

```
Initial LR: 0.1
Step LR: 0.05
Step LR: 0.046000000000000006
Step LR: 0.042
Step LR: 0.038000000000000006
Step LR: 0.034
Step LR: 0.030000000000000002
Step LR: 0.026000000000000002
Step LR: 0.022000000000000002
Step LR: 0.018000000000000002
Step LR: 0.014

```



### Versions

    PyTorch version: 2.12.0+cu130                                                                                                                                                          
    Is debug build: False                                                                                                                                                                  
    CUDA used to build PyTorch: 13.0                                                                                                                                                       
    ROCM used to build PyTorch: N/A                                                                                                                                                        
                                                                                                                                                                                           
    OS: Debian GNU/Linux 12 (bookworm) (x86_64)                                                                                                                                            
    GCC version: (Debian 12.2.0-14+deb12u1) 12.2.0                                                                                                                                         
    Clang version: 14.0.6                                                                                                                                                                  
    CMake version: Could not collect                                                                                                                                                       
    Libc version: glibc-2.36                                                                                                                                                               
                                                                                                                                                                                           
    Python version: 3.11.2 (main, Apr  8 2026, 01:58:00) [GCC 12.2.0] (64-bit runtime)                                                                                                     
    Python platform: Linux-6.1.0-44-cloud-amd64-x86_64-with-glibc2.36                                                                                                                      
    Is CUDA available: False                                                                                                                                                               
    CUDA runtime version: 12.4.131                                                                                                                                                         
    CUDA_MODULE_LOADING set to: N/A                                                                                                                                                        
    GPU models and configuration: Could not collect                                                                                                                                        
    Nvidia driver version: Could not collect                                                                                                                                               
    cuDNN version: Could not collect                                                                                                                                                       
    Is XPU available: False                                                                                                                                                                
    HIP runtime version: N/A                                                                                                                                                               
    MIOpen runtime version: N/A                                                                                                                                                            
    Is XNNPACK available: True                                                                                                                                                             
    Caching allocator config: N/A                                                                                                                                                          
                                                                                                                                                                                           
    CPU:                                                                                                                                                                                   
    Architecture:                            x86_64                                                                                                                                        
    CPU op-mode(s):                          32-bit, 64-bit                                                                                                                                
    Address sizes:                           46 bits physical, 48 bits virtual                                                                                                             
    Byte Order:                              Little Endian                                                                                                                                 
    CPU(s):                                  2                                                                                                                                             
    On-line CPU(s) list:                     0,1                                                                                                                                           
    Vendor ID:                               GenuineIntel                                                                                                                                  
    Model name:                              Intel(R) Xeon(R) CPU @ 2.80GHz                                                                                                                
    CPU family:                              6
    Model:                                   85
    Thread(s) per core:                      1
    Core(s) per socket:                      2
    Socket(s):                               1
    Stepping:                                7
    BogoMIPS:                                5600.32
    Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm          
  constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm   
  abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm avx512f avx512dq rdseed adx smap clflushopt clwb        
  avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat avx512_vnni md_clear arch_capabilities
    Hypervisor vendor:                       KVM
    Virtualization type:                     full
    L1d cache:                               64 KiB (2 instances)
    L1i cache:                               64 KiB (2 instances)
    L2 cache:                                2 MiB (2 instances)
    L3 cache:                                33 MiB (1 instance)
    NUMA node(s):                            1
    NUMA node0 CPU(s):                       0,1
  
    Versions of relevant libraries:
    [pip3] nvidia-cublas==13.1.1.3
    [pip3] nvidia-cuda-cupti==13.0.85
    [pip3] nvidia-cuda-nvrtc==13.0.88
    [pip3] nvidia-cuda-runtime==13.0.96
    [pip3] nvidia-cudnn-cu13==9.20.0.48
    [pip3] nvidia-cufft==12.0.0.61
    [pip3] nvidia-curand==10.4.0.35
    [pip3] nvidia-cusolver==12.0.4.66
    [pip3] nvidia-cusparse==12.6.3.3
    [pip3] nvidia-cusparselt-cu13==0.8.1
    [pip3] nvidia-nccl-cu13==2.29.7
    [pip3] nvidia-nvjitlink==13.0.88
    [pip3] nvidia-nvtx==13.0.85
    [pip3] torch==2.12.0
    [pip3] triton==3.7.0
    [conda] Could not collect


cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

### PR DESCRIPTION
**Pull request by goingforstudying-ctrl**
Fixes #185872

SequentialLR unconditionally initialized the first scheduler and used its learning rate as the initial _last_lr. When milestones contained 0, the second scheduler should already be active at epoch 0, but the first scheduler's LR was used instead, causing all subsequent LRs to be scaled by the wrong start value.

Fix by computing the active scheduler index with bisect_right during __init__, matching the logic already used in step().

### CODE REVIEW

**Review by janeyx99 — COMMENTED** [review]
missing test case!

**Comment by goingforstudying-ctrl** [comment]
I have signed the EasyCLA. The status check may need to be refreshed. Could a maintainer please re-trigger the CLA check?

**Comment by goingforstudying-ctrl** [comment]
EasyCLA is failing. I've already set my git email to goingforstudying-ctrl@users.noreply.github.com. Could you please advise what else is needed to pass the CLA check?

**Comment by goingforstudying-ctrl** [comment]
@janeyx99 Thanks for the review. I've added a test case in test/optim/test_lrscheduler.py that verifies SequentialLR correctly initializes with milestones containing 0. The test checks that the second scheduler is active at epoch 0 when milestones=[0].

**Comment by goingforstudying-ctrl** [comment]
@janeyx99 The missing test case `test_sequentiallr_milestones_contain_zero` has been added in `test/optim/test_lrscheduler.py`. It verifies that `SequentialLR` raises a `ValueError` when `milestones` contains 0. Please let me know if any other changes are needed.

**Comment by janeyx99** [comment]
It looks like you committed 3rd party changes that should not be included. Thnaks for your change but another PR is now landing that fixes the same thing: https://github.com/pytorch/pytorch/pull/185986
